### PR TITLE
PR: Improve and refactor the way we run and debug code in the IPython console

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes

- Depends on https://github.com/spyder-ide/spyder-kernels/pull/286.

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

**Comment by ccordoba12** [comment]
I think we should merge this one before merging PR #15372 because it contains an important refactoring in Spyder-kernels not related to the profiler functionality.

File: spyder/app/tests/test_mainwindow.py [hunk 1]
```diff
@@ -3282,13 +3282,14 @@ def test_debug_unsaved_file(main_window, qtbot):
...
+        qtbot.mouseClick(debug_button, Qt.LeftButton)
 
     # There is a breakpoint, so it should continue
-    qtbot.waitUntil(
-        lambda: '!continue' in shell._control.toPlainText())
-    qtbot.waitUntil(
-        lambda: "1---> 2 print(1)" in control.toPlainText())
+    assert "1---> 2 print(1)" in control.toPlainText()
+    
+    # Verify that we are still  debugging
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py:3291]:
```suggestion
    # Verify that we are still debugging
```

File: spyder/app/tests/test_mainwindow.py [hunk 2]
```diff
@@ -5155,6 +5156,50 @@ def test_add_external_plugins_to_dependencies(main_window, qtbot):
     assert 'spyder-boilerplate' in external_names
 
 
+def test_locals_globals_var_debug(main_window, qtbot, tmpdir):
+    """Test that the debugger can handle variables named globals and locals"""
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py:5160]:
```suggestion
    """Test that the debugger can handle variables named globals and locals."""
```

File: spyder/app/tests/test_mainwindow.py [hunk 3]
```diff
@@ -5155,6 +5156,50 @@ def test_add_external_plugins_to_dependencies(main_window, qtbot):
...
+    ipyconsole = main_window.ipyconsole
+    shell = ipyconsole.get_current_shellwidget()
+    qtbot.waitUntil(lambda: shell._prompt_html is not None,
+                    timeout=SHELL_TIMEOUT)
+    # Test profilefile
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py:5165]:
```suggestion

    # Test code
```

File: spyder/app/tests/test_mainwindow.py [hunk 4]
```diff
@@ -5155,6 +5156,50 @@ def test_add_external_plugins_to_dependencies(main_window, qtbot):
...
+    code = (
+        "globals = 10\n"
+        "def fun():\n"
+        "    locals = 15\n"
+        "    return\n"
+        "fun()"
+        )
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py:5172]:
```suggestion
    )
```

File: spyder/app/tests/test_mainwindow.py [hunk 5]
```diff
@@ -5155,6 +5156,50 @@ def test_add_external_plugins_to_dependencies(main_window, qtbot):
...
+        shell.execute("globals")
+    assert "Out  [3]: 10" in shell._control.toPlainText()
+    with qtbot.waitSignal(shell.executed):
+        shell.execute("locals")
+    assert "Out  [4]: 15" in shell._control.toPlainText()
+
+    with qtbot.waitSignal(shell.executed):
+        shell.execute("q")
+
+    # no errors
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py:5198]:
```suggestion
    # No errors
```

File: spyder/app/tests/test_mainwindow.py [hunk 6]
```diff
@@ -5155,6 +5156,50 @@ def test_add_external_plugins_to_dependencies(main_window, qtbot):
...
+    assert "error" not in shell._control.toPlainText().lower()
+
+
+@pytest.mark.slow
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py:5202]:
```suggestion
```

I remove this marker last month.

File: spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py [hunk 7]
```diff
@@ -1501,18 +1501,20 @@ def test_recursive_pdb(ipyconsole, qtbot):
...
     with qtbot.waitSignal(shell.executed):
         shell.pdb_execute("!debug print()")
-    assert "(IPdb [1]):" in control.toPlainText()
+    assert "((IPdb [1])):" in control.toPlainText()
     with qtbot.waitSignal(shell.executed):
         shell.pdb_execute("!debug print()")
-    assert "((IPdb [1])):" in control.toPlainText()
-    # quit one layer
+    assert "(((IPdb [1]))):" in control.toPlainText()
+    # quit two layer
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py:1515]:
```suggestion
    # Quit two layers
```

File: spyder/plugins/ipythonconsole/widgets/debugging.py [hunk 8]
```diff
@@ -33,14 +34,20 @@
 class SpyderIPy3Lexer(IPython3Lexer):
     # Detect !cmd command and highlight them
     tokens = IPython3Lexer.tokens
-    tokens['root'].insert(
-        0, (r'(!)(\w+)(.*\n)', bygroups(Operator, Keyword, Text)))
+    spyder_tokens = [
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/debugging.py:37]: Why is this change necessary?

File: spyder/plugins/ipythonconsole/widgets/debugging.py [hunk 9]
```diff
@@ -33,14 +34,20 @@
...
+        (r'(!)(\w+)(.*\n)', bygroups(Operator, Keyword, using(Python3Lexer))),
+        (r'(%)(\w+)(.*\n)', bygroups(Operator, Keyword, using(Python3Lexer))),
+        ]
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/debugging.py:40]:
```suggestion
    ]
```

File: spyder/plugins/ipythonconsole/widgets/debugging.py [hunk 10]
```diff
@@ -33,14 +34,20 @@
...
+    tokens['root'] = spyder_tokens + tokens['root']
 
 
 class SpyderIPy2Lexer(IPythonLexer):
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/debugging.py:44]: Should we keep this class? I think it's not necessary anymore because in Spyder 6 it won't be possible to run Python 2 anymore.

File: spyder/plugins/ipythonconsole/widgets/debugging.py (hunk 8 above)
Comment (impact27) [reply spyder/plugins/ipythonconsole/widgets/debugging.py:37]: This adds syntax highlighting for magics which is nice for example:

`%timeit print("a", 0)`

would be highlighted and not left blank

File: spyder/plugins/ipythonconsole/widgets/debugging.py (hunk 10 above)
Comment (impact27) [reply spyder/plugins/ipythonconsole/widgets/debugging.py:44]: Ok if there are no python2 kernels I will remove this

**Comment by impact27** [comment]
> I think we should merge this one before merging PR #15372 because it contains an important refactoring in Spyder-kernels not related to the profiler functionality.

Ok, thanks for the nice review!

**Comment by impact27** [comment]
I have added a test and a fix for a recursive debug issue

File: spyder/app/tests/test_mainwindow.py [hunk 11]
```diff
@@ -6188,5 +6233,58 @@ def test_clickable_ipython_tracebacks(main_window, qtbot, tmpdir):
     qtbot.wait(500)
 
 
+def test_recursive_debug_exception(main_window, qtbot):
+    """
+    Test That an exception in a recursive debug does not break the debugger.
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py:6238]:
```suggestion
    Test that an exception in a recursive debug does not break the debugger.
```

File: spyder/app/tests/test_mainwindow.py [hunk 12]
```diff
@@ -6188,5 +6233,58 @@ def test_clickable_ipython_tracebacks(main_window, qtbot, tmpdir):
...
+    qtbot.waitUntil(
+        lambda: shell.spyder_kernel_ready and shell._prompt_html is not None,
+        timeout=SHELL_TIMEOUT)
+
+    # Main variables
+    shell = main_window.ipyconsole.get_current_shellwidget()
+    control = shell._control
+
+
+    # create new file
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py:6251]:
```suggestion
    # Create new file
```

File: spyder/app/tests/test_mainwindow.py [hunk 13]
```diff
@@ -6188,5 +6233,58 @@ def test_clickable_ipython_tracebacks(main_window, qtbot, tmpdir):
...
+    main_window.editor.new()
+    code_editor = main_window.editor.get_focus_widget()
+    code = 'print("res", 1 + 2)\nprint("res", 2 + 4)'
+    code_editor.set_text(code)
+
+    # debug line
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py:6257]:
```suggestion
    # Debug line
```

File: spyder/app/tests/test_mainwindow.py [hunk 14]
```diff
@@ -6188,5 +6233,58 @@ def test_clickable_ipython_tracebacks(main_window, qtbot, tmpdir):
...
+    with qtbot.waitSignal(shell.executed):
+        shell.execute("debugfile()")
+
+    assert shell.is_debugging()
+    assert '----> 1 print("res", 1 + 2)' in control.toPlainText()
+
+    with qtbot.waitSignal(shell.executed):
+        shell.execute("debug 1/0")
+
+    assert "ENTERING RECURSIVE DEBUGGER" in control.toPlainText()
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py:6267]:
```suggestion
    assert "Entering recursive debugger" in control.toPlainText()
```

To go with my suggestion in https://github.com/spyder-ide/spyder-kernels/pull/286.

File: spyder/app/tests/test_mainwindow.py [hunk 15]
```diff
@@ -6188,5 +6233,58 @@ def test_clickable_ipython_tracebacks(main_window, qtbot, tmpdir):
...
+
+    with qtbot.waitSignal(shell.executed):
+        shell.execute("c")
+
+    assert "ZeroDivisionError" in control.toPlainText()
+    assert "LEAVING RECURSIVE DEBUGGER" in control.toPlainText()
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py:6273]:
```suggestion
    assert "Leaving recursive debugger" in control.toPlainText()
```

**Review by ccordoba12 — COMMENTED** [review]
Some small suggestions for the last test you added, then this should be ready.

**Review by ccordoba12 — APPROVED** [review]
Great work here, thanks @impact27!
